### ISSUE TITLE
default-props-match-prop-types false positive with ...unpacking

### ISSUE THREAD

**Issue by ThiefMaster** [yannickcr/eslint-plugin-react#3138]
```jsx
const testShape = {
  isDeleted: PropTypes.bool,
};

const Foo = ({id, caption, custom, isDeleted}) => {
  console.log(id, caption, custom, isDeleted);
  return <div>test</div>;
};

Foo.propTypes = {
  custom: PropTypes.bool.isRequired,
  ...testShape,
};

Foo.defaultProps = {
  isDeleted: false,
};
```

The `isDeleted: false` line is shown as having an error:

```
defaultProp "isDeleted" has no corresponding propTypes declaration. eslint(react/default-props-match-prop-types)
```

If I remove the `custom` prop from propTypes the error disappears.

eslint v7.32.0
eslint-plugin-react 7.24.0

**Comment by ljharb** [comment]
Dynamically constructing propTypes makes static analysis impossible, except in this exact situation: when the object being spread is statically defined in the same file.

I’d be happy to review a PR that improved propType detection to handle this case, but my suspicion is that you’re much better off explicitly repeating every prop type, both for clarity and for linting.

**Comment by ljharb** [comment]
Closing, but still happy to review such a PR.
